Replace OpenSSL with mbedTLS

Faless opened the pull request:
As discussed at GodotCon this PR **replaces** `OpenSSL` with `mbedTLS`.
Version is `2.7.0`, license is `Apache 2.0`

We lost more than 1MB in the binary size on Linux (gcc, lto, builtin_openssl) which is good.
The whole library size seems to be ~400kb.
```
26940	bin/godot.x11.opt.64    # System OpenSSL on Linux
26928	bin/godot.x11.opt.64    # System mbedTLS on Linux

28660	bin/godot.x11.opt.64    # OpenSSL builtin
27368	bin/godot.x11.opt.64    # mbedTLS builtin
```

I've tested it in multiple ways with this project [SSL_test.zip](https://github.com/godotengine/godot/files/1711990/SSL_test.zip) and by navigating the AssetLib. It seems to work smoothly, but please test it thoroughly.

The module spits an warning on start, complaining that one of the default CA certificates in Godot is invalid.
I have still to figure out which one ( @akien-mga can we maybe try to update it? Not sure how we combine them from the ubuntu source).

The PR also adds a `poll()` function to `StreamPeerSSL` to allow using `get_available_bytes()`.

Building tested for:
- X11 (64 bit)
- OSX (64 bit)
- Windows (64 bit)
- Android (armv7 with neon)
- iPhone

Faless committed:
Add mbedTLS lib. Version 2.7.0, license Apache 2.0

Faless committed:
Deleting OpenSSL module and library

Faless committed:
Disabled mbedtls module in javascript platform

Faless committed:
lws module now uses mbedtls as OpenSSL replacement

Faless committed:
Small change to lws `mbedtls_wrapper` to fix OSX build.

Checks for OSX in thirdparty/lws/mbedtls_wrapper/include/platform/ssl_port.h
to include stdlib.h instead of malloc.h.

The bug is fixed upstream in master via LWS_HAVE_MALLOC_H,
but not in the 2.4.1 branch (as the file structure has, one again,
changed).

Faless committed:
Add mbedtls module

Faless commented:
I like the way GH completly f*** up the history view when you do rebasing.... well... nevermind, `git log` is correct

Calinou commented:
I tested the branch on Fedora 27, it works (asset library loading, I also see the warning message you mentioned). 👍

touilleMan commented:
```
Filed changed 1,106
```

Whoa ! I wouldn't want to be the one reviewing this ! :trollface:

hpvb commented:
That`s great! I haven`t had the chance to look at the whole PR but could you please also allow building against system installed mbedtls? we'd need that to ship in fedora and debian repos.

Faless commented:
> That's great! I haven't had the chance to look at the whole PR but could you please also allow building against system installed mbedtls? we'd need that to ship in fedora and debian repos.

Sure, I'll have a look into it.

Faless committed:
Allow building with system wide mbedtls on X11

Using builtin_mbedtls=yes is still the default as many distributions
do not ship with mbedtls included.

Faless commented:
Should now be able to use system wide mbedtls library by using `builtin_mbedtls=no`.

As state in the comment:
> Using builtin_mbedtls=yes is still the default as many distributions
> do not ship with mbedtls included.

Tested on Ubuntu 17.10, mbedtls version 2.5.1

hpvb commented on platform/x11/detect.py:
```diff
@@ -152,6 +152,10 @@ def configure(env):
     # FIXME: Check for existence of the libs before parsing their flags with pkg-config
 
+    if not env['builtin_mbedtls']:
+        # mbedTLS does not probide a pkgconfig config yet. See https://github.com/ARMmbed/mbedtls/issues/228
```
probide should be provide

akien-mga commented:
Looks like it needs a quick rebase to fix a merge conflict. Will merge afterwards.

akien-mga commented on modules/mbedtls/SCsub:
```diff
@@ -8,84 +8,85 @@ env_mbed_tls = env_modules.Clone()
 # Thirdparty source files
 thirdparty_dir = "#thirdparty/mbedtls/library/"
```
That could go in the `if` too :P (but it's not important)

Faless replied:
Indeed! Done XD

akien-mga commented:
🎉 Thanks a lot Fabio!

rofl0r commented:
good move, yet the points raised in #4267 still apply, as long as the code is embedded in this repo.

akien-mga commented:
> good move, yet the points raised in #4267 still apply, as long as the code is embedded in this repo.

True, but we have no intent to stop vendoring the sources that we need for cross-platform builds, so #4267 won't be fixed.
